### PR TITLE
build: refactor tar code [ch86]

### PR DESCRIPTION
**Pull request by jazzdan**

### CODE REVIEW

File: internal/build/container_updater.go [hunk 1]
```diff
@@ -70,21 +69,19 @@ func (r *containerUpdater) UpdateInContainer(ctx context.Context, cID containerI
...
 func ArchivePathsIfExist(ctx context.Context, paths []pathMapping) (*bytes.Buffer, error) {
-	buf := new(bytes.Buffer)
-	tw := tar.NewWriter(buf)
+	ab := newArchiveBuilder()
 	defer func() {
-		err := tw.Close()
+		err := ab.close()
 		if err != nil {
 			log.Printf("Error closing tar writer: %s", err.Error())
 		}
```
Comment (maiamcc) [inline internal/build/container_updater.go]: why don't we put the error printing in `ab.close()` so we only need to remember to defer one thing?

File: internal/build/tar.go [hunk 2]
```diff
@@ -2,28 +2,46 @@ package build
...
+func (a *archiveBuilder) close() error {
+	return a.tw.Close()
+}
+
+func (a *archiveBuilder) archiveDf(ctx context.Context, df Dockerfile) error {
 	span, ctx := opentracing.StartSpanFromContext(ctx, "daemon-archiveDf")
 	defer span.Finish()
 	tarHeader := &tar.Header{
 		Name: "Dockerfile",
 		Size: int64(len(df)),
```
Comment (maiamcc) [inline internal/build/tar.go]: i bet putting a recent `ModTime` in here will fix that problem with the unrealistic time stamp or whatever

File: internal/build/container_updater.go (hunk 1 above)
Comment (maiamcc) [reply internal/build/container_updater.go]: (tho this is maybe irrelevant if we stop deferring `close()` calls like i'm suggesting in my comment on the PR)

**Comment by jazzdan** [comment]
@maiamcc I agree that it could be better, I'll tackle that in a separate PR now that we have tests!

**Pull request merged**